Sources/SwiftParser/Attributes.swift

```diff
@@ -918,6 +939,75 @@ extension Parser {
   }
 }
 
+extension Parser {
+  mutating func parseABIAttributeArguments(missing: Bool = false) -> RawABIAttributeArgumentsSyntax {
+    let providerDecl: RawABIAttributeArgumentsSyntax.Provider
+    if missing || at(.rightParen) {
+      // There are two situations where the left paren might be missing:
+      //   1. The user just forgot the paren: `@abi var x_abi: Int) var x: Int`
+      //   2. The user forgot the whole argument list: `@abi var x: Int`
+      // Normally, we would distinguish these by seeing if whatever comes after
+      // the attribute parses as an argument. However, for @abi the argument is
+      // a decl, so in #2, we would consume the decl the attribute is attached
+      // to! This leads to a lousy diagnostic in that situation.
+      // Avoid this problem by simply returning a missing decl immediately.
+      // FIXME: Could we look ahead to find an unbalanced parenthesis?
+      providerDecl = .unsupported(
+        RawDeclSyntax(
+          RawMissingDeclSyntax(
+            attributes: self.emptyCollection(RawAttributeListSyntax.self),
+            modifiers: self.emptyCollection(RawDeclModifierListSyntax.self),
+            arena: arena
+          )
+        )
+      )
+    } else {
+      switch nextDeclarationKeyword() {
+      case .group?:
+        providerDecl = RawABIAttributeArgumentsSyntax.Provider(parseDeclarationGroupHeader())!
+
+      case .simple?, .binding?, nil:
+        providerDecl = RawABIAttributeArgumentsSyntax.Provider(parseDeclaration(in: .argumentList))!
```

Comment:
Similar here, I would prefer a solution that doesn’t require force unwrap here because it’s not possible to determine that `RawABIAttributeArgumentsSyntax.Provider` will be non-nil based on local reasoning.

Reply:
We know that it will be non-nil because `parseDeclaration()` returns a `RawDeclSyntax` and `RawABIAttributeArgumentsSyntax.Provider` has a catch-all `unsupported` case which can contain any `RawDeclSyntax`. In other words, the real issue here is that `RawABIAttributeArgumentsSyntax.Provider.init(_:)`.

Would it suffice to add something like this somewhere to limit, document, and encapsulate the force unwrap:

```
extension RawABIAttributeArgumentsSyntax.Provider {
    init(decl: some RawDeclSyntaxNodeProtocol) {
        // Guaranteed to be non-nil because the `unsupported` case which will accept any `RawDeclSyntax`
        self.init(decl)!
    }
}
```

Or do you really think I should add a boilerplate `switch` or `if`-`else` chain solely to avoid a force unwrap? This seems like exactly the situation force unwrap is intended for: The programmer knows for certain that this operation will not be nil, but the type system cannot verify it from the declarations available.

Reply:
I want to be able to do this sort of reasoning locally without needing to look at multiple declarations (because the implementation of `RawABIAttributeArgumentsSyntax.Provider` or `parseDeclaration` might change in the future without re-visiting this invariant. Moving the force unwrap into an initializer of `RawABIAttributeArgumentsSyntax.Provider` seem like a reasonable solution.